swig/cpp/src/Libyang.cpp

```diff
@@ -0,0 +1,297 @@
+    if (NULL == _ctx) {
+        throw runtime_error("can not create new context");
+    }
+    _deleter = S_Deleter(new Deleter(_ctx));
+}
+Context::Context(const char *search_dir, const char *path, LYD_FORMAT format, int options) {
+    _ctx = ly_ctx_new_ylpath(search_dir, path, format, options);
+    if (NULL == _ctx) {
+        throw runtime_error("can not create new context");
+    }
+    _deleter = S_Deleter(new Deleter(_ctx));
+}
+Context::Context(const char *search_dir, LYD_FORMAT format, const char *data, int options) {
+    _ctx = ly_ctx_new_ylmem(search_dir, data, format, options);
+    if (NULL == _ctx) {
+        throw runtime_error("can not create new context");
+    }
+    _deleter = S_Deleter(new Deleter(_ctx));
+}
+Context::~Context() {}
+S_Data_Node Context::info() {
+    struct lyd_node *node = ly_ctx_info(_ctx);
+    return node ? S_Data_Node(new Data_Node(node, _deleter)) : NULL;
+}
+S_Module Context::get_module(const char *name, const char *revision, int implemented) {
+    const struct lys_module *module = ly_ctx_get_module(_ctx, name, revision, implemented);
+    return module ? S_Module(new Module((lys_module *) module, _deleter)) : NULL;
+}
+S_Module Context::get_module_older(S_Module module) {
+    const struct lys_module *new_module = ly_ctx_get_module_older(_ctx, module->_module);
+    return new_module ? S_Module(new Module((lys_module *) new_module, _deleter)) : NULL;
+}
+S_Module Context::load_module(const char *name, const char *revision) {
+    const struct lys_module *module = ly_ctx_load_module(_ctx, name, revision);
+    return module ? S_Module(new Module((lys_module *) module, _deleter)) : NULL;
+}
+S_Module Context::get_module_by_ns(const char *ns, const char *revision, int implemented) {
+    const struct lys_module *module = ly_ctx_get_module_by_ns(_ctx, ns, revision, implemented);
+    return module ? S_Module(new Module((lys_module *) module, _deleter)) : NULL;
+}
+vector<S_Module> *Context::get_module_iter() {
+    const struct lys_module *mod = NULL;
+    uint32_t i = 0;
+
+    auto s_vector = new vector<S_Module>;
+    if (NULL == s_vector) {
+        return NULL;
+    }
+
+    while ((mod = ly_ctx_get_module_iter(_ctx, &i))) {
+        if (mod == NULL) {
+            break;
+        }
+        s_vector->push_back(S_Module(new Module((lys_module *) mod, _deleter)));
+    }
+
+    return s_vector;
+}
+vector<S_Module> *Context::get_disabled_module_iter() {
+    const struct lys_module *mod = NULL;
+    uint32_t i = 0;
+
+    auto s_vector = new vector<S_Module>;
+    if (NULL == s_vector) {
+        return NULL;
+    }
+
+    while ((mod = ly_ctx_get_disabled_module_iter(_ctx, &i))) {
+        if (mod == NULL) {
+            break;
+        }
+        s_vector->push_back(S_Module(new Module((lys_module *) mod, _deleter)));
+    }
+
+    return s_vector;
+}
+void Context::clean() {
+    return ly_ctx_clean(_ctx, NULL);
+}
+vector<string> *Context::get_searchdirs() {
```

Comment:
Why does this return a `std::vector<std::string> *` and not a `std::vector<std::string>`?

Reply:
Had to do it like this because of SWIG, I don't like it either but because of this SWIG cat wrap it into a Python list.

```
folders = ctx.get_searchdirs()
for folder in folders:
    print(folder)
```
Again, sorry about this.